components/skus/browser/rs/lib/src/sdk/credentials/fetch.rs

```diff
@@ -67,6 +79,127 @@ where
+                        num_tokens = num_intervals * num_per_interval;
+                    }
+
+                    let blinded_creds: Vec<BlindedToken> =
+                        match self.client.get_time_limited_v2_creds(&item.id).await? {
+                            Some(item_creds) => {
+                                // we only need new tokens created and signed if we are running low
+                                // so make sure we have less than 5 days of creds before we
+                                // generate more blinded tokens
+                                let valid_to = self
+                                    .last_matching_time_limited_v2_credential(&item.id)
+                                    .await?
+                                    .map(|cred| cred.valid_to)
+                                    .or(valid_to);
+                                if let Some(valid_to) = valid_to {
+                                    // attempt to refresh credentials if we're within 5 days of expiry
+                                    if Utc::now().naive_utc() > (valid_to - chrono::Duration::days(5)) {
+                                        // we have less than 5 days, check if we have any blinded tokens
+                                        // already for this credential
+                                        let tok:Vec<BlindedToken> = item_creds.tokens.iter().map(|t| t.blind()).collect();
+                                        if tok.len() < 1 {
+                                            // we are out of tokens for signing, make some more
+                                            let tokens: Vec<Token> =
+                                                iter::repeat_with(|| Token::random::<Sha512, _>(&mut csprng))
+                                                    .take(num_tokens) // FIXME
+                                                    .collect();
+
+                                            let blinded_creds: Vec<BlindedToken> =
+                                                tokens.iter().map(|t| t.blind()).collect();
+
+                                            self.client.init_time_limited_v2_item_creds(&item.id, tokens).await?;
+                                            blinded_creds
+                                        } else {
+                                            // we have unsigned tokens we need to submit
+                                            tok
+                                        }
+                                    } else {
+                                        // there are enough credentials no need to have any blinded
+                                        // creds created to be signed yet
+                                        Vec::<BlindedToken>::new()
+                                    }
+                                } else {
+                                    // we need to generate some more tokens for this order,
```

Comment:
Ben and I discussed out of band, we need to tweak this logic to ensure that the same batch of creds is used ( if already present ) even if there is no valid_to yet.

Reply:
logic has been updated in the event there are no valid_to credentials yet.